### PR TITLE
[phi] Transfer merged_momentum yaml to phi

### PR DESCRIPTION
**Pull request by jiahy0825**
### PR types
New features

### PR changes
OPs

### Describe
- Transfer `merged_momentum` yaml to phi
- support inplace optional<vector<Tensor>> data structure
- Add function `TensorToConstDenseTensorPtr`, `SetInplaceVectorKernelOutput` and `SetInplaceOptionalVectorKernelOutput`

### CODE REVIEW

File: paddle/phi/api/lib/api_gen_utils.cc [hunk 1]
```diff
@@ -44,6 +44,36 @@ std::unique_ptr<std::vector<phi::DenseTensor*>> TensorToDenseTensor(
   return pt_tensors;
 }
 
+std::vector<const phi::DenseTensor*> TensorToConstDenseTensorPtr(
+    const std::vector<Tensor>& tensors) {
+  std::vector<const phi::DenseTensor*> pt_tensors(tensors.size());
+
+  for (size_t i = 0; i < tensors.size(); ++i) {
+    pt_tensors[i] =
+        std::dynamic_pointer_cast<phi::DenseTensor>(tensors[i].impl()).get();
```
Comment (zyfncg) [inline paddle/phi/api/lib/api_gen_utils.cc]: `dynamic_cast`性能开销会有点多，可以换成`static_cast`

File: paddle/phi/api/lib/api_gen_utils.cc [hunk 2]
```diff
@@ -44,6 +44,36 @@ std::unique_ptr<std::vector<phi::DenseTensor*>> TensorToDenseTensor(
...
+TensorToConstDenseTensorPtr(
+    const paddle::optional<std::vector<Tensor>>& tensors) {
+  paddle::optional<std::vector<const phi::DenseTensor*>> pt_tensors;
+
+  if (tensors) {
+    pt_tensors =
+        paddle::optional<std::vector<const phi::DenseTensor*>>(tensors->size());
+    for (size_t i = 0; i < tensors->size(); ++i) {
+      pt_tensors->at(i) =
+          std::dynamic_pointer_cast<phi::DenseTensor>(tensors->at(i).impl())
```
Comment (zyfncg) [inline paddle/phi/api/lib/api_gen_utils.cc]: 同上

**Review by YuanRisheng — DISMISSED** [review]

**Review by zyfncg — DISMISSED** [review]

File: paddle/phi/api/lib/api_gen_utils.cc (hunk 1 above)
Comment (jiahy0825) [reply paddle/phi/api/lib/api_gen_utils.cc]: Done

File: paddle/phi/api/lib/api_gen_utils.cc (hunk 2 above)
Comment (jiahy0825) [reply paddle/phi/api/lib/api_gen_utils.cc]: Done

**Review by chenwhql — APPROVED** [review]

**Pull request merged**